ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/HiveMetaStoreAuthorizer.java

```diff
@@ -85,38 +97,268 @@ public final void onEvent(PreEventContext preEventContext) throws MetaException,
+    }
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("<== HiveMetaStoreAuthorizer.filterDatabaseObjects() :" + ret );
     }
+    return ret;
+  }
+
+  private List<Table> filterTableObjects(HiveMetaStoreAuthzInfo hiveMetaStoreAuthzInfo, List<Table> tableList) throws MetaException {
+    List<Table> ret = null;
+
+    try {
+      HiveAuthorizer            hiveAuthorizer               = createHiveMetaStoreAuthorizer();
+      List<HivePrivilegeObject> hivePrivilegeObjects         = hiveMetaStoreAuthzInfo.getInputHObjs();
+      HiveAuthzContext          hiveAuthzContext             = hiveMetaStoreAuthzInfo.getHiveAuthzContext();
+      List<HivePrivilegeObject> filteredHivePrivilegeObjects = hiveAuthorizer.filterListCmdObjects(hivePrivilegeObjects, hiveAuthzContext);
+      if (CollectionUtils.isNotEmpty(filteredHivePrivilegeObjects)) {
+        ret = getFilteredTableList(filteredHivePrivilegeObjects, tableList);
+      }
+    } catch (Exception e) {
+      throw new MetaException("Error in HiveMetaStoreAuthorizer.filterTables()" + e.getMessage());
+    }
+    return ret;
+  }
+
+  private List<String> getFilterDatabaseList(List<HivePrivilegeObject> hivePrivilegeObjects) {
+    List<String> ret = new ArrayList<>();
+    for(HivePrivilegeObject hivePrivilegeObject:hivePrivilegeObjects) {
+      String dbName = hivePrivilegeObject.getDbname();
+      ret.add(dbName);
+    }
+    return ret;
+  }
+
+  private List<Table> getFilteredTableList(List<HivePrivilegeObject> hivePrivilegeObjects, List<Table> tableList) {
+    List<Table> ret = new ArrayList<>();
+    for(HivePrivilegeObject hivePrivilegeObject:hivePrivilegeObjects) {
+      String dbName  = hivePrivilegeObject.getDbname();
+      String tblName = hivePrivilegeObject.getObjectName();
+      Table  table   = getFilteredTable(dbName,tblName,tableList);
+      if (table != null) {
+        ret.add(table);
+      }
+    }
+    return ret;
+  }
+
+  private Table getFilteredTable(String dbName, String tblName, List<Table> tableList) {
+    Table ret = null;
+    for (Table table: tableList) {
+      String databaseName = table.getDbName();
+      String tableName = table.getTableName();
+      if (dbName.equals(databaseName) && tblName.equals(tableName)) {
+        ret = table;
+        break;
+      }
+    }
+    return ret;
+  }
+
+  private List<String> filterTableNames(HiveMetaStoreAuthzInfo hiveMetaStoreAuthzInfo, String dbName, List<String> tableNames) throws MetaException {
+    List<String> ret = null;
+
+    try {
+      HiveAuthorizer            hiveAuthorizer               = createHiveMetaStoreAuthorizer();
+      List<HivePrivilegeObject> hivePrivilegeObjects         = hiveMetaStoreAuthzInfo.getInputHObjs();
+      HiveAuthzContext          hiveAuthzContext             = hiveMetaStoreAuthzInfo.getHiveAuthzContext();
+      List<HivePrivilegeObject> filteredHivePrivilegeObjects = hiveAuthorizer.filterListCmdObjects(hivePrivilegeObjects, hiveAuthzContext);
```

Comment:
How big is this list in a worst case? I would be very much afraid to keep all this stuff in memory. I have seen customers with  ~1 million tables. Would it fit into the memory? Even for multiple parallel requests?

Reply:
This shouldn't be too big, because it's per query. All the objects referenced in SQL query gets passed to Ranger plugin to check, so typically the list should not be that big.